Fix bug 1518820: Add errors and warnings

mathjazz opened the pull request:
This is a WIP, see TODO for details. Due to the indent fix, it will be easier to review on a commit-by-commit basis.

TODO:
- [x] Fix showing/hiding of failed checks popup.
- [x] Add ability to `ignore_warnings` (pass param to `update_translation` and add button).
- [x] Close failed checks popup when textarea content changes.

mathjazz commented on frontend/src/modules/editor/actions.js:
```diff
@@ -44,6 +45,30 @@ export function updateSelection(content: string): UpdateSelectionAction {
+    +pErrors: Array<string>,
+    +clWarnings: Array<string>,
+    +pndbWarnings: Array<string>,
+    +ttWarnings: Array<string>,
```
I know, I know... I looked into refactoring checks before digging into it, but the patch would be pretty big. Maybe I should file a P5 Translate.Next bug, so we don't forget about it?

I'm thinking checks should instead be formatted like this:
```js
failedChecks = {
    errors: [
        {
            message: 'Missing variable %s',
            library: 'compare-locales'
        },
        ...
    ],
    warnings: [...],
}
```

adngdb replied:
Oh god yes that looks so much better! What I was wondering is if you could maybe add a layer between the API and the reducer to make the data from the server look more like what you show up there? It might be too much work, though, feel free not to do that.

However, yes, please do file a bug about refactoring errors and warnings. :-) Not a Translate.Next though, I think it's a general Pontoon issue.

mathjazz replied:
I'm thinking the bug I'm about to file would make such API redundant, so I'd rather not spent time working on it.

mathjazz replied:
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1545740.

@jotes

jotes replied:
@mathjazz thanks! I totally agree.  The current structure of the API response is a bit complicated (I didn't consider the later use and scaling). Your proposal looks very good :-)

mathjazz commented on frontend/src/modules/editor/components/Editor.js:
```diff
@@ -100,6 +101,27 @@ export class EditorBase extends React.Component<InternalProps> {
+        const plural = pluralForm === -1 ? 0 : pluralForm;
+        const translation = selectedEntity.translation[plural];
+
+        if (!translation || (!translation.approved && !translation.fuzzy)) {
```
We only show checks of the active translation if it's approved or fuzzy (because than its status is also colored as error/warning in the string list).

But now that I think about it, maybe we should just show the failed checks by default anyways? Otherwise the user might think it's OK to approve the translation, and once she approves it, she'll see the popup (which is not an optimal UX).

What do you think?

adngdb replied:
That sounds reasonable indeed, though maybe we should refrain from changing the behavior from Translate.Current for now?

mathjazz replied:
Good point! I'll add a comment.

mathjazz commented on frontend/src/modules/editor/components/FailedChecks.js:
```diff
@@ -0,0 +1,65 @@
+                    className="close"
+                    onClick={ this.closeFailedChecks }
+                >×</button>
+            </Localized>
```
@stasm Is there a way to make only the `ariaLabel` localized, without the button text?

stasm replied:
Sadly, not right now. There's an open issue about it: https://github.com/projectfluent/fluent.js/issues/184

mathjazz replied:
Thanks, that's a good enough answer right now.

mathjazz commented on frontend/src/modules/editor/components/Editor.js:
```diff
@@ -101,6 +107,34 @@ export class EditorBase extends React.Component<InternalProps> {
+            state.router,
+            ignoreWarnings,
+        ));
+    }
```
This is very ugly, because `updateTranslationStatus` already exists in `EntityDetails` component, but I can't pass it to the `Editor`.

mathjazz commented:
@adngdb I've addressed your feedback and made two additional changes:
* Failed checks are now always stored in the editor (including the ones coming from the DB). That allows us to use `resetFailedChecks` to close the failed checks popup.
* Save/submit/approve Anyway button is now available.

Could you have another look, please?

There's one open issue remaining - the failed checks popup should close when you change the content of the editor. If I `resetFailedChecks` in `Editor.componentDidUpdate`, I hit the infinite recursion of course. Is the way to go here to factor `FailedChecks` out into their own component?

I wanted to hear your opinion first, because there might be other refactorings needed here. See my [comment](https://github.com/mozilla/pontoon/pull/1266#discussion_r276540834). Thanks!

adngdb commented on frontend/src/modules/editor/reducer.js:
```diff
@@ -32,7 +32,13 @@ export type EditorState = {|
     +warnings: Array<string>,
-    +source: '' | 'stored' | 'submitted' | 'approved',
+
+    // A source of failed checks (erros and warnings). Possible values:
```
s/erros/errors

mathjazz commented:
> Regarding the panel closing issue, can you maybe show me a commit with the problem you hit? It might just be a matter of finding the right condition.

Pushed https://github.com/mozilla/pontoon/pull/1266/commits/dc1a3fbb1bd90c6c258b6d550ce23952076e1646. You should see the following error in the console on any translate.next view:
```
Error: Maximum update depth exceeded.
This can happen when a component repeatedly calls setState inside
componentWillUpdate or componentDidUpdate.
React limits the number of nested updates to prevent infinite loops.
```

mathjazz commented:
That prevents failed checks from appearing... ever.

The reason for this is that we show/hide failed checks by updating the **`editor`** state. So when we add checks in `updateFailedChecks`, `editor` state changes, forcing `GenericEditor` component to re-render, because it relies on the `editor` state. After render, `componentDidUpdate` is called, which then calls `resetFailedChecks` (because the condition is `true`).

That's why I'm thinking, maybe we should make `FailedChecks` a separate module, with its state independent of the editor?

mathjazz commented:
> It seems your patch breaks the stats in the FilterPanel. This happens after I change the status of a translation (approve, reject):

That bug actually happens on master, too.

adngdb commented:
I've tried locally, and this does the trick:

```js
        if (
            (
                prevProps.editor.errors.length ||
                prevProps.editor.warnings.length
            ) && (
                this.props.editor.errors.length ||
                this.props.editor.warnings.length
            )
        ) {
            this.props.resetFailedChecks();
        }
```

You don't want to reset if the errors and warnings were just added to the editor object, but in all other cases you do.

mathjazz commented:
> I've tried locally, and this does the trick:

Thanks, that is super helpful! I've added that, now I'll write some tests and then I'll ask you for another round of review.

mathjazz commented:
@adngdb This is now ready for another look. I can't figure out why one of the tests is failing (left a [comment](https://github.com/mozilla/pontoon/pull/1266/commits/d44c772eb21aa12975ec8226bcfd1fc9eba3cc39) in the commit message). I'd appreciate if you could enlighten me. :)

adngdb commented on frontend/src/modules/entitydetails/components/EntityDetails.test.js:
```diff
@@ -107,6 +107,43 @@ describe('<EntityDetailsBase>', () => {
+            },
+        });
+
+        expect(editor.actions.updateFailedChecks.calledOnce.toBeTruthy());
```
```js
expect(editor.actions.updateFailedChecks.calledOnce).toBeTruthy();
```

`toBeTruthy` is a property of the `expect` framework. :-)

adngdb replied:
For this to work, you'll need to mock those two actions though. See for example: https://github.com/mozilla/pontoon/blob/master/frontend/src/modules/entitydetails/components/EntityDetails.test.js#L115

mathjazz replied:
Thanks :) Updated, but the 2nd test is still failing.

adngdb commented on frontend/src/modules/editor/components/EditorProxy.test.js:
```diff
@@ -24,4 +31,41 @@ describe('<EditorProxy>', () => {
+    it('resets failed checks when translation changes, but errors and warnings do not', () => {
+        const resetMock = sinon.stub();
+
+        const wrapper = mount(<EditorProxy
```
I'm not sure you need to `mount` here, `shallow` might be sufficient. It's better to avoid using `mount` as much as possible, because `mount` will instantiate all the components in the hierarchy below this one, whether `shallow` only instantiate the one component you're testing, and no others.

adngdb commented on frontend/src/modules/editor/components/EditorProxy.test.js:
```diff
@@ -24,4 +31,41 @@ describe('<EditorProxy>', () => {
+    it('keeps failed checks when translation changes, along with errors or warnings', () => {
+        const resetMock = sinon.stub();
+
+        const wrapper = mount(<EditorProxy
```
Same here.

adngdb commented on frontend/src/modules/entitydetails/components/EntityDetails.test.js:
```diff
@@ -107,6 +117,45 @@ describe('<EntityDetailsBase>', () => {
+        });
+
+        expect(editor.actions.updateFailedChecks.calledOnce).toBeFalsy();
+        expect(editor.actions.resetFailedChecks.calledOnce).toBeTruthy();
```
Have you checked that the method that calls this action is actually called? I usually run those kinds of debugs with `console.log`, and make sure that the correct path gets ran.

mathjazz replied:
Thanks for the debugging help. :) Indeed, there was a problem with the code and I fixed it.

But I'm only **almost** there... For some reason `editor.actions.updateFailedChecks.calledOnce` is still `truthy`. I thought count should be reset in `afterAll`?

adngdb replied:
Ah, `afterAll` only runs after all tests have run. You want to use `afterEach` as well, see: https://github.com/mozilla/pontoon/blob/master/frontend/src/modules/editor/components/Editor.test.js#L66

adngdb commented on frontend/src/modules/entitydetails/components/EntityDetails.test.js:
```diff
@@ -96,12 +96,12 @@ function createEntityDetailsWithStore() {
     });
 
-    afterAll(() => {
+    afterEach(() => {
```
I think that's the wrong approach. Those were fine. I believe what you want to do is add a third step, `afterEach`, that calls `.reset()`. Otherwise, you mock and de-mock the object after every test. I have no idea if that has an impact in terms of performance, but I'm pretty sure that's not the philosophy of `sinon`.

mathjazz replied:
I tried that first, but it gave errors:
* `TypeError: editor.actions.updateFailedChecks.restore is not a function`
* `Received: undefined`

adngdb replied:
I've just tested locally with this code and it passes: 

```js
    beforeAll(() => {
        sinon.stub(editor.actions, 'updateFailedChecks').returns({ type: 'whatever'});
        sinon.stub(editor.actions, 'resetFailedChecks').returns({ type: 'whatever'});
    });

    afterEach(() => {
        editor.actions.updateFailedChecks.reset();
        editor.actions.resetFailedChecks.reset();
    });

    afterAll(() => {
        editor.actions.updateFailedChecks.restore();
        editor.actions.resetFailedChecks.restore();
    });
```

mathjazz commented:
Thanks for the advice, the tests should now work as expected!
